### PR TITLE
DAOS-4282 object: verify the specified shard for fetch

### PR DESCRIPTION
**Pull request by Nasf-Fan**
Sometimes, we need to fetch data from specified shard, such as
for verifying replicas consistency. Under such case, the shard
corresponding to the dkey_hash should match the given shard.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### CODE REVIEW

**Comment by Nasf-Fan** [comment]
To be retested because of DAOS-4285.

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -1618,8 +1623,18 @@ obj_req_get_tgts(struct dc_object *obj, enum obj_rpc_opc opc, int *shard,
...
+				    *shard >= rc + obj->cob_grp_size) {
+					D_ERROR("Fetch from invalid shard, "
+						"grp size %u, shard cnt %u, "
+						"grp idx %u, given shard %u, "
+						"dkey hash %lu, dkey %s\n",
+						obj->cob_grp_size,
+						obj->cob_shards_nr,
+						rc / obj->cob_grp_size, *shard,
+						dkey_hash,
+						(char *)dkey->iov_buf);
```
Comment (liuxuezhao) [inline src/object/cli_obj.c]: some dkey is not printable, and some cases may pass NULL dkey(like in dc_obj_sync)?

Comment (Nasf-Fan) [reply src/object/cli_obj.c]: It is only for DAOS_OBJ_RPC_FETCH case.

**Comment by Nasf-Fan** [comment]
> Just curious, did you see this failure in racer?

The issue can be reproduced by "daos_test -V". It is reported by both Liwei and Samir Raval. But I did not see failure instance during daos_racer test.

**Comment by Nasf-Fan** [comment]
Test failed because of DAOS-4285.

**Comment by Nasf-Fan** [comment]
Test failed because of DAOS-4285.

**Comment by Nasf-Fan** [comment]
I would suggest to force land the patch.

**Pull request merged**